InnerEye/ML/reports/classification_multilabel_report.py

```diff
@@ -0,0 +1,173 @@
+from pathlib import Path
+from typing import List, Set, FrozenSet
+
+import pandas as pd
+import torch
+import math
+
+from InnerEye.Common.metrics_constants import LoggingColumns
+from InnerEye.ML.dataset.scalar_dataset import ScalarDataset
+from InnerEye.ML.scalar_config import ScalarModelBase
+from InnerEye.ML.reports.classification_report import LabelsAndPredictions, print_metrics, get_labels_and_predictions, \
+    get_metric, ReportedMetrics
+from InnerEye.ML.reports.notebook_report import print_header
+
+
+def get_unique_prediction_target_combinations(config: ScalarModelBase) -> Set[FrozenSet[str]]:
+    """
+    Get a list of all the combinations of labels that exist in the dataset.
+
+    For multilabel classification tasks, this function will return all unique combinations of labels that
+    occur in the dataset csv.
+    For example, if there are 6 samples in the dataset with the following ground truth labels
+    Sample1: class1, class2
+    Sample2: class0
+    Sample3: class1
+    Sample4: class2, class3
+    Sample5: (all label classes are negative in Sample 5)
+    Sample6: class1, class2
+    This function will return {{"class1", "class2"}, {"class0"}, {"class1"},  {"class2", "class3"}, {}}
+
+    For binary classification tasks (assume class_names has not been changed from ["Default"]):
+    This function will return a set with two members - {{"Default"}, {}} if there is at least one positive example
+    in the dataset. If there are no positive examples, it returns {{}}.
+    """
+    df = config.read_dataset_if_needed()
+    dataset = ScalarDataset(args=config, data_frame=df)
+
+    all_labels = [torch.flatten(torch.nonzero(item.label)).tolist() for item in dataset.items]
+    label_set = set(frozenset([config.class_names[i] for i in labels if not math.isnan(i)])
+                    for labels in all_labels)
+
+    return label_set
+
+
+def get_dataframe_with_exact_label_matches(metrics_df: pd.DataFrame,
+                                           prediction_target_set_to_match: List[str],
+                                           all_prediction_targets: List[str],
+                                           thresholds_per_prediction_target: List[float]) -> pd.DataFrame:
+    """
+    Given a set of prediction targets, for each sample find
+        (i) if the set of ground truth labels matches this set exactly,
+        (ii) if the predicted model outputs (after thresholding) match this set exactly
+
+    Generates an output dataframe with the rows:
+    LoggingColumns.Patient, LoggingColumns.Label, LoggingColumns.ModelOutput, LoggingColumns.Hue
+
+    The output dataframe is generated according to the following rules:
+      - LoggingColumns.Patient: For each sample, the sample id is copied over into this field
+      - LoggingColumns.Label: For each sample, this field is set to 1 if the set of ground truth labels for the sample
```

Comment:
Am really confused was it the difference between set of ground truth labels and set of prediction targets?